examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -170,6 +148,43 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
         latch.countDown();
     }
 
+    private int processMessages()
+        throws ProducerFencedException, FencedInstanceIdException {
+        try {
+            ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofMillis(200));
+            if (records.count() > 0) {
+                // Begin a new transaction session.
+                producer.beginTransaction();
+                for (ConsumerRecord<Integer, String> record : records) {
+                    // Process the record and send to downstream.
+                    ProducerRecord<Integer, String> customizedRecord = transform(record);
+                    producer.send(customizedRecord);
+                }
+                Map<TopicPartition, OffsetAndMetadata> positions = new HashMap<>();
+                for (TopicPartition topicPartition : consumer.assignment()) {
+                    positions.put(topicPartition, new OffsetAndMetadata(consumer.position(topicPartition), null));
+                }
+                // Checkpoint the progress by sending offsets to group coordinator broker.
+                // Under group mode, we must apply consumer group metadata for proper fencing.
+                if (this.mode.equals("groupMode")) {
+                    producer.sendOffsetsToTransaction(positions, consumer.groupMetadata());
+                } else {
+                    producer.sendOffsetsToTransaction(positions, consumerGroupId);
+                }
+
+                // Finish the transaction. All sent records should be visible for consumption now.
+                producer.commitTransaction();
+                return records.count();
+            }
+        } catch (CommitFailedException | WakeupException | InterruptException | TimeoutException e) {
```

Comment:
I think we can leave `WakeupException` and `InterruptException` out of this. In both of these cases, we would probably just want the application to close. I think the main thing we want this example to show is the "normal operating" exceptions.

Reply:
Sounds good